Apache HttpClient 4.5 for WGet HTTP transport

ok2c opened the pull request.

mickaelistria commented on src/main/java/com/googlecode/download/maven/plugin/internal/HttpFileRequester.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright 2009-2016 Marc-Andre Houle and Red Hat Inc
```
If you created this class, please make sure the Copyright shows the right date and the right owner (you or your employer depending on your agreement)

ok2c replied:
@mickaelistria OK. Corrected.

mickaelistria replied:
Are you working for the Apache Foundation? If not, you're not supposed to cease your copyright.

You should either put your name or the name of your company or employer, depending on the context you're doing this. It could be something like `Copyright (c) 2018, OK2Consulting`.

ok2c replied:
@ok2c I am contributing this code as an ASF member, not as an employee of OK2 Consulting.

mickaelistria replied:
Ok, I do not know the terms of Apache Foundation regarding copyrights (are members allow to copyright their work on the name of the Foundation?). I know that for Eclipse Foundation, members/committers have to use their own or employer copyright and are not permitted to set Copyright to Eclipse Foundation.
If you confirm me that it's a practice that's accepted by the Apache Foundation, then I'm fine with it and will resolve.

ok2c replied:
@ok2c I am not a legal expert, but as per CLA I have with ASF I can grant copyright of my code contributions and IP therewith to ASF.

mickaelistria replied:
Ok, thanks!

mickaelistria commented:
@longtimeago would you like to have a look?

longtimeago commented:
@mickaelistria sure, why not :)

longtimeago commented on src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java:
```diff
@@ -332,50 +371,81 @@ private void unpack(File outputFile) throws NoSuchArchiverException {
-            authenticationInfo.setUserName(username);
-            authenticationInfo.setPassword(password);
+
+            credentialsProvider.setCredentials(
```
@ok2c we could create here UsernamePasswordCredentials only and then create and configure `BasicCredentialsProvider` only ones after `if` block

longtimeago commented on src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java:
```diff
@@ -332,50 +371,81 @@ private void unpack(File outputFile) throws NoSuchArchiverException {
         }
 
-        AuthenticationInfo authenticationInfo = new AuthenticationInfo();
+        final BasicCredentialsProvider credentialsProvider = new BasicCredentialsProvider();
```
@ok2c let's use interface `CredentialsProvider` for the field type instead of `BasicCredentialsProvider`

longtimeago commented on src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java:
```diff
@@ -332,50 +371,81 @@ private void unpack(File outputFile) throws NoSuchArchiverException {
+
+        final CloseableHttpClient httpClient = HttpClientBuilder.create()
+                .setConnectionManager(CONN_POOL)
+                .setConnectionManagerShared(true)
```
@ok2c afaiu, setting this to `true` requires the pull to be shutdowned explicitly, no?

ok2c replied:
@longtimeago yes, it does. Given that the connection pool is static, we want to make sure it does not get shut down by mistake.

longtimeago commented on src/main/java/com/googlecode/download/maven/plugin/internal/WGet.java:
```diff
@@ -332,50 +371,81 @@ private void unpack(File outputFile) throws NoSuchArchiverException {
+
+        final HttpFileRequester fileRequester = new HttpFileRequester(
+                httpClient,
+                this.session.getSettings().isInteractiveMode() ? getLog() : null);
```
@ok2c instead of passing null and then check for null everywhere we introduce `SilentLog` class implementing `org.apache.maven.plugin.logging.Log`, which will simply do nothing.

This of course would imply that byte calculations are thrown away for non-interactive mode but let's avoid premature optimizations - readability is our first goal :)

longtimeago commented on src/main/java/com/googlecode/download/maven/plugin/internal/HttpFileRequester.java:
```diff
@@ -0,0 +1,103 @@
+
+                    } catch (IOException ex) {
+                        if (log != null) {
+                            log.error(ex);
```
@ok2c bot sure we want to suppress IO errors in non-interactive mode

longtimeago commented on src/main/java/com/googlecode/download/maven/plugin/internal/HttpFileRequester.java:
```diff
@@ -0,0 +1,103 @@
+                            out.write(tmp, 0, bytesRead);
+
+                            if (log != null) {
+                                completed += bytesRead;
```
@ok2c I would suggest to move all byte logging stuff to a separate class to don't mess it and byte reading/writing

E.g. `ProgressBar` with methods 
`add(int bytesRead)` - printing one intermediate results 
`done()` - printing final result

longtimeago commented on src/main/java/com/googlecode/download/maven/plugin/internal/HttpFileRequester.java:
```diff
@@ -0,0 +1,103 @@
+                    long completed = 0L;
+
+                    byte[] tmp = new byte[8 * (int) KBYTE];
+                    try (InputStream in = entity.getContent(); FileOutputStream out = new FileOutputStream(outputFile)) {
```
@ok2c we should wrap FileOutputStream with buffered version to avoid hitting OS with every singe byte.
Also, please, use `OutputStream` as `out` variable type

ok2c replied:
@longtimeago Why? There is no single byte write operation anywhere, unless you want to entertain a sceanrio of target server sending a single byte in an IP packet.

longtimeago replied:
@ok2c indeed, sorry, missed the fact that we are writing from a buffer

longtimeago commented on src/main/java/com/googlecode/download/maven/plugin/internal/HttpFileRequester.java:
```diff
@@ -0,0 +1,103 @@
+
+    public void download(final URI uri, final File outputFile, final HttpContext clientContext) throws Exception {
+        final HttpGet httpGet = new HttpGet(uri);
+        httpClient.execute(httpGet, new ResponseHandler<Void>() {
```
@ok2c not a fan of anonymous classes of such size. Let's extract it to named inner class

ok2c commented:
@longtimeago I will fix things that are functionally wrong. As far as stylistic preferences are concerned you you are welcome to refactor the code as you deem appropriate.

ok2c commented:
@longtimeago I refactored progress feedback logic per your suggestion and also implemented most of the changes you have requested. I am not at all sure that on-demand instantiation of `BasicCredentialsProvider` made anything better other than saving one object allocation on the heap.

ok2c commented:
Feel free to suggest better interface / class and method names.

ok2c commented:
@longtimeago Added javadocs. Please review.

mickaelistria commented:
Thanks guys!
@longtimeago I let you merge this (pick the strategy you prefer, squashing or not, I don't really mind here)
